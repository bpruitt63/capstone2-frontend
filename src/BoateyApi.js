import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL;


class BoateyApi {

    static token;

    /** Sends request to database and catches any errors */
    static async request(endpoint, data={}, method="get") {
        const url = `${BASE_URL}/${endpoint}`;
        const headers = {Authorization: `Bearer ${BoateyApi.token}`};
        const params = (method === "get") ? data : {};

        try {
            return (await axios({url, method, data, params, headers})).data;
        } catch (err) {
            console.error("API Error:", err.response);
            let message;
            !err.response ? message = "Server error, please try again later"
                : message = err.response.data.error.message;
            throw Array.isArray(message) ? message : [message];
        };
    };

    /** Register new user and get JWT */
    static async register(data) {
        const res = await this.request('users/register', data, 'post');
        return res.token;
    };

    /** Log in existing user and get JWT. 
     * Returns user location and preferred unit of measurement.
     */
    static async login(data) {
        const res = await this.request('users/login', data, 'post');
        const token = res.token;
        const location = res.location;
        const units = res.units;
        return {token, location, units};
    };

    /** Gets all info on a single user */
    static async getUser(username) {
        const res = await this.request(`users/${username}`);
        return res.user;
    };

    /** Updates selected data on a single user and returns all data */
    static async updateUser(username, data) {
        const res = await this.request(`users/${username}`, data, 'patch');
        return res.user;
    };

    /** Adds location into database if not already added */
    static async addLocation(data) {
        const res = await this.request(`locations/new_location`, data, 'post');
        return res.location;
    };

    /** Adds location rating into database */
    static async addRating(username, data) {
        const res = await this.request(`locations/rating/${username}`, data, 'patch');
        return res.rating;
    };

    /** Gets user and overall ratings for a point */
    static async getRatings(username, id) {
        const res = await this.request(`locations/${id}/${username}`);
        const ratings = {overallRating: res.overallRating, 
                        userRating: res.userRating};
        return ratings;
    };

    /** Adds trip and its locations into database */
    static async addTrip(username, data) {
        const res = await this.request(`trips/${username}/new_trip`, data, 'post');
        const trip = {trip: res.trip, tripLocations: res.tripLocations};
        return trip;
    };

    /** Gets logged in user's trips from database */
    static async getUserTrips(username) {
        const res = await this.request(`trips/${username}/trips`);
        return res;
    };

    /** Gets 20 most recent trips from database */
    static async getRecentTrips() {
        const res = await this.request('trips/recent_trips');
        return res;
    };

    /** Deletes trip from database */
    static async deleteTrip(username, id) {
        const res = await this.request(`trips/${username}/${id}/delete`, {}, 'delete');
        return res;
    };

    /** Adds trip rating to database */
    static async rateTrip(username, id, data) {
        const res = await this.request(`trips/${username}/${id}/rate`, data, 'patch');
        return res;
    };
};

BoateyApi.token = localStorage.getItem("token");

export default BoateyApi;