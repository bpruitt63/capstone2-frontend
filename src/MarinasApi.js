import axios from 'axios';
//import {marinas} from './ApiKeys';

const BASE_URL = 'https://api.marinas.com/v1/';
const API_KEY = process.env.REACT_APP_MARINAS_API_KEY;

class MarinasApi {

    /** Retrieves array of all points of interest within a certain distance
     * of a specific latitude/longitude.
     */
    static async getPoints(longLat) {
        try {
            const res = await axios.get(
                `${BASE_URL}points/search?location[lon]=${longLat[0]}&location[lat]=${longLat[1]}&location[radius]=20000&access_key=${API_KEY}`
            );
            return res;
        } catch(err) {
            console.error("marinas api Error:", err.response);
            let message;
            !err.response ? message = "Server error, please try again later"
                : message = err.response.data.error.message;
            throw Array.isArray(message) ? message : [message];
        };
    };

    /** Retrieves info on a specific point of interest */
    static async getOnePoint(id) {
        try {
            const res = await axios.get(
                `${BASE_URL}points/${id}`
            );
            return res.data;
        } catch(err) {
            console.error("marinas api Error:", err.response);
            let message;
            !err.response ? message = "Server error, please try again later"
                : message = err.response.data.error.message;
            throw Array.isArray(message) ? message : [message];
        };
    };
};

export default MarinasApi;