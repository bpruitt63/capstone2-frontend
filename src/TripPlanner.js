import React, {useState, useEffect} from 'react';
import {Link} from 'react-router-dom';
import MarinasApi from './MarinasApi';
import Point from './Point';
import CurrentTrip from './CurrentTrip';
import TripsList from './TripsList';
import BoateyApi from './BoateyApi';
import WeatherApi from './WeatherApi';

/** Top level component for trip planning feature.  Shows all points
 * of interest near specified location, and allows logged in users to 
 * create their own trips.
 */
function TripPlanner({longLat, units, username, setWeather}) {

    const [points, setPoints] = useState();
    const [currentTrip, setCurrentTrip] = useState(JSON.parse(localStorage.getItem("currentTrip")) || []);
    const [userTrips, setUserTrips] = useState();
    const [recentTrips, setRecentTrips] = useState();
    const [showRecentTrips, setShowRecentTrips] = useState(false);
    const [showUserTrips, setShowUserTrips] = useState(false);
    const isSaved = false;

    useEffect(() => {

        /** Retrieves array of nearby points and sets in state */
        async function getPoints() {
            try {
                const p = await MarinasApi.getPoints(longLat);
                setPoints(p.data.data);
            } catch (e) {
                setPoints('API error');
            };
        };
        getPoints();
    }, [longLat, setPoints]);

    useEffect(() => {

        /** Retrieves all of logged in user's saved trips */
        async function getUserTrips() {
            if (username){
                try {
                    const trips = await BoateyApi.getUserTrips(username);
                    setUserTrips(trips);
                } catch (e) {
                    setUserTrips("Can't retrieve trip information");
                };
            };
        };
        getUserTrips();
    }, [username, setUserTrips]);

    useEffect(() => {

        /** Retrieves last saved trips from database */
        async function getRecentTrips() {
            try {
                const trips = await BoateyApi.getRecentTrips();
                setRecentTrips(trips);
            } catch (e) {
                setRecentTrips("Can't retrieve trip information");
            };
        };
        getRecentTrips();
    }, [setRecentTrips]);

    /** Removes all current trip data */
    const deleteCurrentTrip = () => {
        setCurrentTrip([]);
        localStorage.removeItem("currentTrip");
    };

    /** Toggles visibility of recent trips */
    const toggleShowRecent = () => {
        setShowRecentTrips(!showRecentTrips);
    };

    /** Toggles visibility of user's trips */
    const toggleShowUser = () => {
        setShowUserTrips(!showUserTrips);
    };

    /** Sets weather and points list to center on the longitude/latitude
     * of the selected point.
     */
    const searchHere = async (coordinates) => {
        try {
            const w = await WeatherApi.getWeatherbyLongLat(coordinates, units);
            setWeather(w);
        } catch (e) {
            setWeather('API error');
        };
        try {
            const p = await MarinasApi.getPoints(coordinates);
            setPoints(p.data.data);
        } catch (e) {
            setPoints('API error');
        };
    };

    /** TODO adds newly created trip to saved/recent trips lists */
    // const addedTrip = (trip) => {
    //     setUserTrips([trip, ...userTrips]);
    // };


    /** Deletes trip from database.  
     * 
     * TODO removes trip from saved/recent trips lists
     */
    const deleteTrip = async (username, tripId) => {
        await BoateyApi.deleteTrip(username, tripId);
        //const trips = [...userTrips];
        const trips = [];
        for (let trip of userTrips) {
            trips.push({...trip});
        };
        const removed = trips.findIndex(function(trip) {
            return trip.tripId === tripId;
        });
        trips.splice(removed, 1);
        setUserTrips(trips);
    };

    return (
        <div>
            {!username && 
                <p>
                    <Link to='/login'>Login</Link> or{' '}
                    <Link to='/register'>register</Link>{' '}
                    to create your own trips.
                </p>
            }   
            {showRecentTrips && 
                <h3>Recently Created Trips</h3>}
            {showRecentTrips && recentTrips && 
                <TripsList trips={recentTrips}
                            units={units} />}
            <button onClick={toggleShowRecent}>
                {showRecentTrips ? 'Hide Recent Trips' : 'Show Recent Trips'}
            </button>
            {username &&
                <div>
                    {userTrips && showUserTrips &&
                        <>
                            <h3>Your Saved Trips</h3>
                            <TripsList trips={userTrips}
                                       units={units}
                                       username={username}
                                       deleteTrip={deleteTrip} />
                        </>
                    }
                    <button onClick={toggleShowUser}>
                        {showUserTrips ? 'Hide Saved Trips' : 'Show Saved Trips'}
                    </button>
                    <CurrentTrip currentTrip={currentTrip}
                            deleteCurrentTrip={deleteCurrentTrip}
                            units={units}
                            username={username} />
                </div>
            }
            <h3>Nearby Points of Interest</h3>
            {!points ? <p>Loading...</p>
                : points === 'API error' ? <p>Error retrieving marina information.  Please try again later</p>
                : !points[0] ? <p>No points of interest nearby</p> 
                : points.map(p => 
                <Point point={p}
                        key={p.id}
                        currentTrip={currentTrip}
                        setCurrentTrip={setCurrentTrip}
                        username={username}
                        isSaved={isSaved}
                        units={units}
                        searchHere={searchHere} />
            )}
        </div>
    );
};

export default TripPlanner;