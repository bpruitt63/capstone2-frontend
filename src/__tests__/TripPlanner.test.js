import React from 'react';
import { render, screen, fireEvent, waitForElementToBeRemoved } from '@testing-library/react';
import TripPlanner from '../TripPlanner';
import { BrowserRouter } from 'react-router-dom';
import MarinasApi from '../MarinasApi';
import BoateyApi from '../BoateyApi';


jest.mock('mapbox-gl/dist/mapbox-gl', () => ({
    Map: () => ({}),
}));  

jest.mock('../MarinasApi');

jest.mock('../BoateyApi');

const marinasRes = {data: {data: [{id: 'test1', 
                                    kind: 'marina',
                                    name: 'Test Location',
                                    review_count: 0,
                                    web_url: 'https://google.com',
                                    location: {lat: 40, lon: 10},
                                    fuel: {has_diesel: false,
                                            has_propane: false,
                                            has_gas: true}},
                                    {id: 'test2', 
                                    kind: 'ramp',
                                    name: 'Test Location 2',
                                    review_count: 0,
                                    web_url: 'https://google.com',
                                    location: {lat: 50, lon: 15},
                                    fuel: {has_diesel: false,
                                            has_propane: false,
                                            has_gas: false}}]}};

const tripsRes = [{tripId: 1, 
                    tripName: 'test1', 
                    distance: '1000', 
                    tripRating: 1, 
                    locations: [{locationId: "testLoc1", 
                                locationPosition: 0, 
                                locationName: 'testLocName1'},
                                {locationId: "testLoc2", 
                                locationPosition: 1, 
                                locationName: 'testLocName2'}]},
                    {tripId: 2, 
                    tripName: 'test2', 
                    distance: '1000', 
                    tripRating: null, 
                    locations: [{locationId: "testLoc2", 
                                locationPosition: 0, 
                                locationName: 'testLocName2'},
                                {locationId: "testLoc1", 
                                locationPosition: 1, 
                                locationName: 'testLocName1'}]}];

test('renders', async () => {
    MarinasApi.getPoints.mockResolvedValue(marinasRes);
    BoateyApi.getUserTrips.mockResolvedValue(tripsRes);
    BoateyApi.getRecentTrips.mockResolvedValue(tripsRes);
    render(<BrowserRouter><TripPlanner/></BrowserRouter>);
    await waitForElementToBeRemoved(() => screen.getByText(/Loading.../));
});

test('matches snapshot', async () => {
    MarinasApi.getPoints.mockResolvedValue(marinasRes);
    BoateyApi.getUserTrips.mockResolvedValue(tripsRes);
    BoateyApi.getRecentTrips.mockResolvedValue(tripsRes);
    const {asFragment} = render(<BrowserRouter><TripPlanner/></BrowserRouter>);
    await waitForElementToBeRemoved(() => screen.getByText(/Loading.../));
    expect(asFragment()).toMatchSnapshot();
});

test('renders logged out', async () => {
    MarinasApi.getPoints.mockResolvedValue(marinasRes);
    BoateyApi.getUserTrips.mockResolvedValue(tripsRes);
    BoateyApi.getRecentTrips.mockResolvedValue(tripsRes);
    render(<BrowserRouter><TripPlanner/></BrowserRouter>);
    await waitForElementToBeRemoved(() => screen.getByText(/Loading.../));
    expect(screen.getByText(/to create your own trips./)).toBeInTheDocument;
    expect(screen.queryByText(/Show Saved Trips/)).not.toBeInTheDocument;
    expect(screen.getByText(/Test Location 2/)).toBeInTheDocument;
});

test('renders logged in', async () => {
    MarinasApi.getPoints.mockResolvedValue(marinasRes);
    BoateyApi.getUserTrips.mockResolvedValue(tripsRes);
    BoateyApi.getRecentTrips.mockResolvedValue(tripsRes);
    render(<BrowserRouter><TripPlanner username={'test'}/></BrowserRouter>);
    await waitForElementToBeRemoved(() => screen.getByText(/Loading.../));
    expect(screen.queryByText(/to create your own trips./)).not.toBeInTheDocument;
    expect(screen.getByText(/Show Saved Trips/)).toBeInTheDocument;
    expect(screen.getByText(/Test Location 2/)).toBeInTheDocument;
});

test('toggles shown trips logged out', async () => {
    MarinasApi.getPoints.mockResolvedValue(marinasRes);
    BoateyApi.getUserTrips.mockResolvedValue(tripsRes);
    BoateyApi.getRecentTrips.mockResolvedValue(tripsRes);
    render(<BrowserRouter><TripPlanner/></BrowserRouter>);
    await waitForElementToBeRemoved(() => screen.getByText(/Loading.../));
    expect(screen.queryByText(/test1/)).not.toBeInTheDocument;
    expect(screen.queryByText(/Show Saved Trips/)).not.toBeInTheDocument;
    fireEvent.click(screen.getByText(/Show Recent Trips/));
    expect(screen.getByText(/test1/)).toBeInTheDocument;
    fireEvent.click(screen.getByText(/Hide Recent Trips/));
    expect(screen.queryByText(/test1/)).not.toBeInTheDocument;
});

test('toggles shown trips logged in', async () => {
    MarinasApi.getPoints.mockResolvedValue(marinasRes);
    BoateyApi.getUserTrips.mockResolvedValue(tripsRes);
    BoateyApi.getRecentTrips.mockResolvedValue(tripsRes);
    render(<BrowserRouter><TripPlanner username={'test'}/></BrowserRouter>);
    await waitForElementToBeRemoved(() => screen.getByText(/Loading.../));
    expect(screen.queryByText(/test1/)).not.toBeInTheDocument;
    fireEvent.click(screen.getByText(/Show Saved Trips/));
    expect(screen.getByText(/test1/)).toBeInTheDocument;
    fireEvent.click(screen.getByText(/Hide Saved Trips/));
    expect(screen.queryByText(/test1/)).not.toBeInTheDocument;
});