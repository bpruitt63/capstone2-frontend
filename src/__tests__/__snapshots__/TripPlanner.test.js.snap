// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <p
      class="links"
    >
      <a
        href="/login"
      >
        Login
      </a>
       or 
      <a
        href="/register"
      >
        register
      </a>
       to create your own trips.
    </p>
    <br />
    <div
      class="notUsed"
    >
      <br />
      <button
        class="toggle btn btn-secondary"
        type="button"
      >
        Show Recent Trips
      </button>
    </div>
    <div
      class="points"
    >
      <h3
        class="pointsTitle"
      >
        Nearby Points of Interest
      </h3>
      <div
        class="searchedPoint"
      >
        <h4
          class="pointTitle"
        >
          Test Location
        </h4>
        <p>
          <span
            class="detailTitle"
          >
            Location type: 
          </span>
          Marina
        </p>
        <div
          class="collapse"
        />
        <button
          class="pointButton btn btn-secondary"
          type="button"
        >
          Center Search Here
        </button>
        <button
          class="pointButton btn btn-secondary"
          type="button"
        >
          Show Details
        </button>
      </div>
      <div
        class="searchedPoint"
      >
        <h4
          class="pointTitle"
        >
          Test Location 2
        </h4>
        <p>
          <span
            class="detailTitle"
          >
            Location type: 
          </span>
          Ramp
        </p>
        <div
          class="collapse"
        />
        <button
          class="pointButton btn btn-secondary"
          type="button"
        >
          Center Search Here
        </button>
        <button
          class="pointButton btn btn-secondary"
          type="button"
        >
          Show Details
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
